diag_manager/fms_diag_yaml.F90

```diff
@@ -564,19 +555,41 @@ subroutine diag_get_value_from_key(diag_file_id, par_id, key_name, value_name, i
 end subroutine diag_get_value_from_key
 
 !> @brief gets the lat/lon of the sub region to use in a diag_table yaml
-subroutine get_sub_region(diag_yaml_id, sub_region_id, sub_region)
-  integer, intent(in)  :: diag_yaml_id       !< Id of the diag_table yaml file
-  integer, intent(in)  :: sub_region_id      !< Id of the region block to read from
-  class(*),intent(out) :: sub_region (NUM_SUB_REGION_ARRAY) !< Array storing the bounds of the sub region
-
-  call get_value_from_key(diag_yaml_id, sub_region_id, "dim1_begin", sub_region(1), is_optional=.true.)
-  call get_value_from_key(diag_yaml_id, sub_region_id, "dim1_end", sub_region(2), is_optional=.true.)
-  call get_value_from_key(diag_yaml_id, sub_region_id, "dim2_begin", sub_region(3), is_optional=.true.)
-  call get_value_from_key(diag_yaml_id, sub_region_id, "dim2_end", sub_region(4), is_optional=.true.)
-  call get_value_from_key(diag_yaml_id, sub_region_id, "dim3_begin", sub_region(5), is_optional=.true.)
-  call get_value_from_key(diag_yaml_id, sub_region_id, "dim3_end", sub_region(6), is_optional=.true.)
-  call get_value_from_key(diag_yaml_id, sub_region_id, "dim4_begin", sub_region(7), is_optional=.true.)
-  call get_value_from_key(diag_yaml_id, sub_region_id, "dim4_end", sub_region(8), is_optional=.true.)
+subroutine get_sub_region(diag_yaml_id, sub_region_id, sub_region, fname)
+  integer,             intent(in)    :: diag_yaml_id       !< Id of the diag_table yaml file
+  integer,             intent(in)    :: sub_region_id      !< Id of the region block to read from
+  type(subRegion_type),intent(inout) :: sub_region         !< Type that stores the sub_region
+  character(len=*),    intent(in)    :: fname              !< filename of the subregion (for error messages)
+
+  select case (trim(sub_region%grid_type))
+  case ("latlon")
+    allocate(real(kind=r4_kind) :: sub_region%corner1(2))
+    allocate(real(kind=r4_kind) :: sub_region%corner2(2))
+    allocate(real(kind=r4_kind) :: sub_region%corner3(2))
+    allocate(real(kind=r4_kind) :: sub_region%corner4(2))
+  case ("index")
```

Comment:
The "latlon" and "index" should be parameter variables instead of hardcoded values.